Bundle operands to specify denormal modes

spavloff opened the pull request:
Two new operands are now supported in the "fp.control" operand bundle:

    * "denorm.in=xxx"   - specifies the inpot denormal mode.
    * "denorm.out=xxx"  - specifies the output denormal mode.

Here xxx must be one of the following values:

    * "ieee"  - preserve denormals.
    * "zero"  - flush to zero preserving sign.
    * "pzero" - flush to positive zero.
    * "dyn"   - mode is dynamically read from a control register.

These values align those permitted in the "denormal-fp-math" function attribute.

spavloff committed:
Bundle operands to specify denormal modes

Two new operands are now supported in the "fp.control" operand bundle:

    * "denorm.in=xxx"   - specifies the inpot denormal mode.
    * "denorm.out=xxx"  - specifies the output denormal mode.

Here xxx must be one of the following values:

    * "ieee"  - preserve denormals.
    * "zero"  - flush to zero preserving sign.
    * "pzero" - flush to positive zero.
    * "dyn"   - mode is dynamically read from a control register.

These values align those permitted in the "denormal-fp-math" function
attribute.

arsenm commented on llvm/lib/IR/Instructions.cpp:
```diff
@@ -678,6 +682,71 @@ fp::ExceptionBehavior CallBase::getExceptionBehavior() const {
+    return DenormalMode::IEEE;
+
+  Type *Ty = nullptr;
+  for (auto &A : args())
+    if (auto *T = A.get()->getType(); T->isFPOrFPVectorTy()) {
+      Ty = T;
+      break;
+    }
+  assert(Ty && "Some input argument must be of floating-point type");
```
This probably can't be an assert 

What happens if there are multiple input floating point type arguments, and each one has a differently controllable denormal mode?

For example, AMDGPU has some instructions with mixed half and float inputs, but those read from different FP mode controls (I haven't actually checked the right input mode bits are respected in these cases).

I'd probably just stop trying to report the mode if there are multiple input types

spavloff replied:
It means that we have more than one denormal mode in a single instruction, so they need to be distinguish them somehow, by type in your case.

Is it possible that an instruction has multiple output denormal modes?

What about rounding mode? It it possible that evaluation of an instruction requires two different rounding modes?

spavloff replied:
Introduced "denorm.f32.in=" and "denorm.f32.out=" bundle operands.

arsenm commented on llvm/lib/IR/Instructions.cpp:
```diff
@@ -678,6 +682,71 @@ fp::ExceptionBehavior CallBase::getExceptionBehavior() const {
+    }
+  }
+
+  if (!getParent())
+    return DenormalMode::IEEE;
+  const Function *F = getFunction();
+  if (!F)
+    return DenormalMode::IEEE;
```
I don't think this low level query should try to get creative and inspect the calling context function. It should stick to just parsing the bundle

spavloff replied:
Removed this search.

arsenm reviewed (commented):
Ho

spavloff committed:
Address review comments

- introduced bundle operands for 32-bit vslues,
- simplified demormal mode search,
- updated documentation.

spavloff committed:
Fix some errors
